### ISSUE TITLE
letrec* incorrectly reorders statements

### ISSUE THREAD

**Issue by fare** [gambit/gambit#659]
R7RS says that `letrec*` should not reorder the side-effects, that will happen from left-to-right, but this function, when compiled (as opposed to interpreted, which works correctly), reorders line 300 below line 500, which is incorrect. Presumably, it detects that 400 and 500 do not "depend" on x (or anything else defined in the `letrec*`), whereas 300 does.

```
":" ; ( echo -e 'INTERPRETED' ; gsi letrec-star-reordered.scm; echo -e '\nCOMPILED' ; gsc letrec-star-reordered.scm && gsi -e '(load "letrec-star-reordered")' ) 2>&1 | less

(define (displayln . a) (for-each display a) (newline))

(define (reordered)
  (letrec* ((x (begin (displayln 100) 42))
            (_200 (displayln 200))
            (_300 (displayln 300 (list x)))
            (_400 (displayln 400))
            (y (begin (displayln 500) 23)))
    (displayln 600 (list y))))

(reordered)
```

**Comment by fare** [comment]
Marc said on gitter: Concerning the implementation of `letrec*`, there are two implementations… a correct one in the interpreter (procedure `##comp-let-like-form` in `lib/_eval.scm`) and a buggy one in the compiler (procedure `pt-recursive-let` in `gsc/_ptree1.scm`).

**Comment by gambiteer** [comment]
Fixed by 47f2eb7feb5deae94c0b96f5a50310d978e120ae

**Comment by gambiteer** [comment]
Sometimes we make mistakes.  Reopening.

**Comment by fare** [comment]
NB: the fix was immediately reverted in the next commit d1a6c6bf5ea019731c4264ed7a91c0386659a989 because it broke some regression tests... that I believe might need be updated.

**Comment by SamuelYvon** [comment]
@fare 
```shell
$ ./gsc/gsc
Gambit v4.9.4-211-g97e4bff5

> (define (displayln . a) (for-each display a) (newline))
> (define (reordered)
   (letrec* ((x (begin (displayln 100) 42))
            (_200 (displayln 200))
            (_300 (displayln 300 (list x)))
            (_400 (displayln 400))
            (y (begin (displayln 500) 23)))
    (displayln 600 (list y))))
> (reordered)

100
200
300(42)
400
500
600(23)
```

**Comment by fare** [comment]
Yes, this is the desired behaviour. As of 4.9.4-210-g420f1e5e, it works in the interpreted case, and fails in the compiled case with
```
100
200
400
500
300(42)
600(23)
```

**Comment by SamuelYvon** [comment]
Do you have a link to the spec where conditions about reordering can be found?

**Comment by fare** [comment]
I suppose that would be r7rs section 4.2.2 "Binding constructs": https://standards.scheme.org/corrected-r7rs/r7rs-Z-H-6.html#TAG:__tex2page_sec_4.2.2

**Comment by fare** [comment]
Why would that be illegal? You bind `x` then you bind `y` to `x`. What would be illegal is trying to bind `y` to `x` before `x` was bound.

**Comment by SamuelYvon** [comment]
It was too early in the morning, my example is miswritten, I was trying to construct something that refers to:

> If it is not possible to evaluate each <init> without assigning or referring to the value of the corresponding <variable> or the <variable> of any of the bindings that follow it in 
<bindings>, it is an error

from the R7RS spec. I flipping the `<init>` would be a valid example of that.

Something like this "breaks" in chicken scheme:

```scheme
(letrec* ((x y)
          (y (lambda (z) (+ z 5))))
         (print x))
```

**Comment by fare** [comment]
Agreed. (Except I make no comment on the implementation, that I didn't look at properly). `letrec*` might not need any topological sort at all.

**Comment by feeley** [comment]
I have just discovered that commit 2420124 has introduced a major loss of optimization of internal defines, in particular when they are used to define constants. This happens in various places in the runtime system, such as the numerical library and `equal?-hash` that is critical to the performance of `table` lookups (I have measured a 5x slowdown in `table-ref` because of this commit).

As an example the following code:

```
(define (foo)
  (define const 42)
  const)
```

which should be treated like `(define (foo) 42)` (and was treated like that before the commit) is treated like:

```
(define (foo)
  (let ((const (box #!void)))
    (set-box! const 42)
    (unbox const)))
```

Needless to say this prevents many optimizations, such as constant propagation and constant folding and it creates pressure on the garbage collector because it needlessly allocates boxes.

I am working on a fix.
